stream/transport/redis/redis.go

```diff
@@ -2,42 +2,85 @@ package stream
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
+	"time"
 
+	"github.com/blocknative/dreamboat/stream/transport"
+	"github.com/google/uuid"
+	"github.com/jpillora/backoff"
 	"github.com/lthibault/log"
 	"github.com/redis/go-redis/v9"
 )
 
-type Pubsub struct {
-	Redis  *redis.Client
-	Logger log.Logger
+type Topic struct {
+	Redis     *redis.Client
+	Logger    log.Logger
+	Name      string
+	LocalNode uuid.UUID
 }
 
-func (r *Pubsub) Publish(ctx context.Context, topic string, data []byte) error {
-	return r.Redis.Publish(ctx, topic, data).Err()
+func (t *Topic) Publish(ctx context.Context, m transport.Message) error {
+	m.Source = t.LocalNode
+
+	b, err := json.Marshal(&m)
+	if err != nil {
+		return fmt.Errorf("marshal json: %w", err)
+	}
+
+	return t.Redis.Publish(ctx, t.Name, b).Err()
+}
+
+func (t *Topic) Subscribe(ctx context.Context) transport.Subscription {
+	return Subscription{
+		PubSub:    t.Redis.Subscribe(ctx, t.Name),
+		LocalNode: t.LocalNode,
+	}
+}
+
+type Subscription struct {
+	*redis.PubSub
+	Logger    log.Logger
+	LocalNode uuid.UUID
 }
 
-func (r *Pubsub) Subscribe(ctx context.Context, topic string) chan []byte {
-	logger := r.Logger.WithField("topic", topic)
-
-	sub := make(chan []byte)
-	go func() {
-		defer close(sub)
-		for ctx.Err() == nil { // restart on failure
-			pubsub := r.Redis.Subscribe(ctx, topic)
-			logger.Debug("redis subscription started")
-
-			redisSub := pubsub.Channel()
-			for data := range redisSub {
-				select {
-				case sub <- []byte(data.Payload):
-				case <-ctx.Done():
-					return
-				}
-			}
-			logger.Warn("redis subscription closed")
+func (s Subscription) Next(ctx context.Context) (transport.Message, error) {
+	var b = backoff.Backoff{
+		Min:    time.Millisecond,
+		Max:    time.Millisecond * 100,
+		Jitter: true,
+	}
+
+	for {
+		rawMsg, err := s.ReceiveMessage(ctx)
+		switch err {
+		case redis.ErrClosed, context.Canceled:
+			return transport.Message{}, err
+		}
+
+		if err != nil {
+			s.Logger.
+				WithError(err).
+				WithField("attempt", b.Attempt()).
+				WithField("backoff", b.ForAttempt(b.Attempt())).
+				Warn("failed to get subscription message from redis")
+			time.Sleep(b.Duration())
+			continue
 		}
 
-	}()
+		b.Reset()
 
-	return sub
+		msg, err := transport.Decode([]byte(rawMsg.Payload))
```

Comment:
In the previous implementation decoding was done in multiple goroutines so that we've been decoding messages in parallel. That's why subscribe was separated from the rest

Reply:
Yeah, I noticed that.  I'm happy to re-introduce this in a follow-up PR, but this one was already getting quite large.  I also suspect that the parallelism isn't needed here, but maybe you've already measured this?

Reply:
For the parallelism -  It may not be for bids. for payloads however especially that you discard own payloads after the publish it may be signifficant change. As it's up to 300kB per message to decode and we'd know it would grow soon in deneb to +500kB

Reply:
> for payloads however especially that you discard own payloads after the publish it may be signifficant change. 

Just to make sure we're tracking:  we discard payloads that originate from the local node, but the payload doesn't need to be decoded to make this determination.  All that needs to happen is for the envelope (`transport.Message`) to be decoded; the payload is decoded separately.

Therefore, it seems like this would not be very expensive. (Right?)